ENH: Speed up trim_zeros

BvB93 opened the pull request:
Closes https://github.com/numpy/numpy/issues/16783.

As was noted in aforementioned issue the current `trim_zeros()` implementation, 
as of 1.19.0, is very slow with plenty of room for further optimization. 
This pull request addresses the previous optimization issue. 

Before
-------
``` python
In [1]: import numpy as np

In [2]: a = np.hstack([
   ...:     np.zeros(100_000),
   ...:     np.random.uniform(size=100_000),
   ...:     np.zeros(100_000),
   ...: ])

In [3]: np.__version__
Out[3]: '1.19.0'

In [4]: %timeit np.trim_zeros(a)
45.8 ms ± 6.38 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
```

After
-----
``` python
In [3]: np.__version__
Out[3]: '1.20.0.dev0+2823c98'

In [4]: %timeit np.trim_zeros(a)
303 µs ± 15 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

To summarize the new implementation: the passed array is now first converted a boolean array,
after which `argmax()` is used for identifying the first and/or last non-zero element.

~~A side effect of the new approach is that it will trim any leading and/or trailing elements
which evaluates to ``False``; not just ``0``.~~

Lastly, a new benchmark has been added as was requested in https://github.com/numpy/numpy/issues/16783#issuecomment-655856960.

Commit:
Added a benchmark for `trim_zeros()`

Commit:
Updated the benchmark with `param_names` and `params`

Commit:
Improve the performance of `np.trim_zeros()`

Commit:
Fixed a string-representation

Commit:
docstring update; fixed an issue with the try/except approach in 'B'

Commit:
Removed a sentence which is redundant since 2a2d9d622e9bdeafd5cfafc9f3d20fd8d27c235f

Commit:
Reverted a redundant documentation change

Commit:
Improve readability: use ndarray.any() to check for empty arrays

Commit:
Revert "Improve readability: use ndarray.any() to check for empty arrays"

This reverts commit 997bdc793234d6c1c75738035ba63730d686b356.

Commit:
The performance increase of 997bdc793234d6c1c75738035ba63730d686b356 is significant for large arrays; revert it

Commit:
Increase the variety of the tests

BvB93 commented on numpy/lib/function_base.py:
```diff
@@ -1616,27 +1619,45 @@ def trim_zeros(filt, trim='fb'):
-                break
-            else:
-                last = last - 1
+    try:
```
Without such ``try`` / ``except`` approach you get weird situation were one can pass a list of strings but not an array of strings:

``` python

In [1]: import numpy as np                                                                                                                             

In [2]: a = ['a', 'b', 'c', '', 'b', 'a']                                                                                              

In [3]: np.asanyarray(a, dtype=bool)                                                                                                                   
Out[3]: 
array([True,  True,  True, False,  True,  True])

In [4]: b = np.array(a)                                                                                                                                

In [5]: np.asanyarray(b, dtype=bool)                                                                                                                   
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
    ...
ValueError: invalid literal for int() with base 10: ''
```

Commit:
BUG: check if array-length is non-zero before calling `argmax()`

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1616,27 +1619,47 @@ def trim_zeros(filt, trim='fb'):
+        arr_any = np.asanyarray(filt)
+        _arr = arr_any.view(bool)
+        _arr.shape = arr_any.shape + (arr_any.dtype.itemsize,)
+        arr = _arr.any(axis=-1)
```
I don't think this branch makes sense - I think you want either:
```
    arr = asanyarray(filt) == 0
```
or
```
    arr = asanyarray(filt).astype(bool, copy=False)
```

BvB93 replied:
I actually hadn't even considered that; implemented in https://github.com/numpy/numpy/pull/16911/commits/bb21711f862ad5d1212af727440b617d38f8c73b.

eric-wieser reviewed (commented):
I hadn't realized that `trim_zeros` wasn't written for arrays at all.

I worry though that we can't change it, else code like `trim_zeros([[1], 0])` will stop working.

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1616,27 +1619,47 @@ def trim_zeros(filt, trim='fb'):
+    if 'B' in trim_upper:
+        last = len_arr - arr[::-1].argmax()
+        # If `last == len(arr) and arr[-1] is False` then all elements are False
+        if last == len_arr and not arr[-1]:
```
```suggestion
        if not arr[-1]:
```
The only way it can be false is if it is last

BvB93 replied:
Note that `arr` is not trimmed (yet) at this point, so `arr[-1] is False` is very much a possibility without all other elements being `False`.

eric-wieser replied:
Ah, sorry, What I meant is
```suggestion
        if not arr[last]:
```

BvB93 replied:
Ah yes, that's definitely much clearer; implemented in https://github.com/numpy/numpy/pull/16911/commits/6ea881c1d9c956378ffc2d6445126ac485f45d04.

Note that ``arr[last - 1]`` had to be used here as you'd otherwise get an IndexError if ``last == len(arr)``.

BvB93 commented:
> I hadn't realized that `trim_zeros` wasn't written for arrays at all.

Yeah, it's most definitely a bit of an oddball.

> I worry though that we can't change it, else code like `trim_zeros([[1], 0])` will stop working.

Object arrays should work fine as long as they can be either converted into boolean arrays (such as is the case with your example).
``` python
In [1]: import numpy as np                                                                                                                             

In [2]: np.array([[1], 0], dtype=object).astype(bool)                                                                                                  
Out[2]: array([ True, False])
```

The only exception I can think of is a (ragged) object array consisting of other arrays, 
though these didn't work with the old implementation either is it involved array to scalar comparisons.

``` python
In [1]: import numpy as np

In [2]: a = np.empty(2, dtype=object)

In [3]: a[0] = a[1] = np.random.rand(5,)

In [4]: a.astype(bool)
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
    ...
ValueError: setting an array element with a sequence.```

BvB93 commented:
Circleci failure seems to be caused by the unrelated `histogram2d()` documentation.

eric-wieser commented:
> Object arrays should work fine as long as they can be either converted into boolean arrays (such as is the case with your example).

Right, but your example code only works because you explicitly added `dtype=object`. From what I can tell, this fails / emits a deprecation warning with your code;
```python
>>> np.trim_zeros([[1], 0])
[[1]]
```
and this fails outright:
```python
>>> np.trim_zeros([[1]])
[[1]]
```

Commit:
Implement https://github.com/numpy/numpy/pull/16911#discussion_r457330789

Commit:
Implement https://github.com/numpy/numpy/pull/16911#discussion_r457358202

eric-wieser commented on numpy/lib/function_base.py:
```diff
@@ -1622,21 +1622,29 @@ def trim_zeros(filt, trim='fb'):
+        last = len_arr - arr[::-1].argmax()
+        # If `arr[last - 1] is False` then so are all other elements
+        if not arr[last - 1]:
+            return filt[len_arr:]
```
```suggestion
            return filt[:0]
```
would be slightly truer to intent here, "trim the back of the array"

BvB93 replied:
I agree; implemented in https://github.com/numpy/numpy/pull/16911/commits/d75061d6cbf32575a30dfe3628d38e073a9220d4.

Commit:
Implemented https://github.com/numpy/numpy/pull/16911#discussion_r457415065

seberg commented:
As silly as `trim_zeros` is, I think the reason it uses a for loop, is probably speed to begin with: It is probably *much* faster as is if you only trim a handfull of zeros (or even none).  However, its plausible that you can argue it does not matter:  Its like having a fast-path for `np.all()` when the first element is already `False`:  It saves an arbitrary amount of time, but if you do any _other_ operation on the array later, the speedup is probably dwarfed by that.

BvB93 commented:
> As silly as `trim_zeros` is, I think the reason it uses a for loop, is probably speed to begin with: It is probably much faster ~~as is~~ if you only trim a handfull of zeros (or even none). 

Indeed, the example execution times I provided are for a pretty large array; 
the difference will be much smaller for when the input array is smaller.

Fortunetly, even arrays as small as `np.ones(1)` show a minor improvements in execution time (~4 µs (new) vs ~4.5 µs (old)).

BvB93 commented:
> Fortunetly, even arrays as small as `np.ones(1)` show a minor improvements in execution time (~4 µs (new) vs ~4.5 µs (old)).

On second thought, I suspect this is moreso related to the old implementation's lack of `enumerate()` and the `if i == 0.:` rather than `if i:` comparisons

BvB93 commented:
> and this fails outright:
> ``` python
> >>> np.trim_zeros([[1]])
> [[1]]
> ```

After giving it a bit more thought there might be a very simple way of dealing with this issue: 
simply return the passed `filt` if it cannot be coerced into a 1D array (instead of raising a `ValueError`).

While I feel raising an error would generally speaking be more appropriate, 
this alternative solution is definitely safer as far as backwards compatibility is concerned.

eric-wieser commented:
> Maybe in combination with a deprecation warning of some sort?

Perhaps the thing to do is something like:
```
if isinstance(arr, np.ndarray):
    # your optimal code
else:
   # emit deprecation warning
   # old code
```

Another more cynical view would be just deprecate this function entirely as it's trivial slow and unlike most of numpy, and work out a better name / submodule / PyPI package for the fast `trim_zeros`.

Commit:
Implemented https://github.com/numpy/numpy/pull/16911#issuecomment-661231106

Fall back to the old `np.trim_zeros()` implementation if an exception is encountered.
Emit a `DeprecationWarning` in such case.

BvB93 commented:
> Perhaps the thing to do is something like:
> ```python
> ...
> ```

I just added something like this in https://github.com/numpy/numpy/pull/16911/commits/a6f9d296e79c8a1de65bfd18a17bdb7b1ed427c7.
It will now automatically fall back to the old implementation if an exception is encountered (where the same exception may or may not be raised again, depending on its nature.

> Another more cynical view would be just deprecate this function entirely as it's trivial slow and unlike most of numpy, and work out a better name / submodule / PyPI package for the fast `trim_zeros`.

If the option was just between leaving it in and removing it then I would agree with this view, 
as in its current state It's IMO not really something that belongs in numpy's public API.
However as `trim_zeros()` still seems salvageable, I'm personally leaning more towards leaving it in.

jonashaag commented on numpy/lib/function_base.py:
```diff
@@ -1621,6 +1621,56 @@ def trim_zeros(filt, trim='fb'):
+    except Exception as ex:
+        warning = DeprecationWarning(
+            "in the future trim_zeros will require a 1-D array as input"
+            "which is compatible with ndarray.astype(bool)"
```
I don’t want to be a smart ass, but this is important.

The sentence should use “that” rather than “which”. The difference is that with “which” the sentence says that all 1-D arrays are compatible with astype(bool), but I guess you meant to say that the array must be 1-D AND compatible with astype(bool).

It’s a common mistake made by German natives, and probably also Dutch (from your name I assume you are Dutch).

matthew-brett replied:
And English natives!  I often make that mistake myself...

BvB93 replied:
> The sentence should use “that” rather than “which”. The difference is that with “which” the sentence says that all 1-D arrays are compatible with astype(bool), but I guess you meant to say that the array must be 1-D AND compatible with astype(bool).

Out of curiosity, isn't there a distinction between cases where "which" is preceded by a comma ("... a 1-D array as input, which is compatible ...") versus ones without comma ("... a 1-D array as input which is compatible ...")? 
The first scenario definitely conveys the incorrect meaning, but I was under the impression that whether or not the second one is acceptable is more of a American vs British thing.

In any case, it should be changed for the sake of clarity (see https://github.com/numpy/numpy/pull/16911/commits/4ddfc3eb68202a626d8bb318963679f8e831b1e6).

jonashaag replied:
I *guess* there is no distinction because English is "insensitive" to commas in many cases (and also British and American place commas differently). But that's just a guess and I'd be interested in a definitive answer if anybody knows it :-)

matthew-brett replied:
I think you're right, that the 'which' there makes it a little more likely that someone could get the impression that any 1D array can be cast as bool.  I agree that the impression is less using 'that'.   But like you, I couldn't give you a definitive source for that (or which, oh no, so confusing).

Commit:
TST: Avoid the use of `warnings.simplefilter('ignore', ...)

Commit:
MAINT: Gramar fix

Commit:
TST,MAINT: Fixed an incorrect exception name

rossbar commented:
Just to note: the current CI failures are unrelated. A rebase on master should fix them.

mattip commented:
New deprecations require
-  [ ] a date/version comment like [this one](https://github.com/numpy/numpy/blob/master/numpy/core/_methods.py#L96),
- [ ] tests that the warning works in [test/test_deprecations](https://github.com/numpy/numpy/blob/master/numpy/core/tests/test_deprecations.py) and
- [ ] a [release note](https://github.com/numpy/numpy/tree/master/doc/release/upcoming_changes).

Commit:
STY: remove trailing whitespaces

Commit:
DEP: Added a date/version deprecation comment

Commit:
MAINT: Typo fix; added a missing space

Commit:
TST,DEP: Added a deprecation testcase

Commit:
DEP,REL: Added a deprecation release note

BvB93 commented:
Thanks @mattip! 
Your comments have, respectively, been addressed in https://github.com/numpy/numpy/pull/16911/commits/853b83a6dee09d190a0ee9caa099a5b8a1f8cd94, https://github.com/numpy/numpy/pull/16911/commits/65f5634808a925837c2ec41116a0b12ed162f510 and https://github.com/numpy/numpy/pull/16911/commits/d531cd7f735a47039a9ffc88493c8837c8d2a2fd.

As a side note, the branch has been rebased to get rid of a merge conflict.

mattip commented on numpy/lib/tests/test_function_base.py:
```diff
@@ -1166,25 +1166,63 @@ def test_subclass(self):
+                raise AssertionError('Failed to raise an exception')
+
+        out = nfb._trim_zeros_old(arr)
+        assert_allclose(out, arr)
```
This test can be removed since it now is in test_deprecations.

BvB93 replied:
Done as of https://github.com/numpy/numpy/pull/16911/commits/823fe8b60dd3e4bd10f66237ef59771c17d53c00; all deprecation related tests are now in ``numpy/core/tests/test_deprecations.py``.

Commit:
TST,DEP: Moved leftovers from a deprecation test

BvB93 commented:
Still failing due to the ``setuptools`` issue.
Rebasing in order to apply https://github.com/numpy/numpy/pull/16993.

mattip commented:
Thanks @BvB93

eric-wieser commented:
I think this breaks `np.trim_zeros([2**64, 0])`. It should probably either mention this in the release notes, or we should emit a deprecation warning on all non-array inputs.

mattip commented:
Both before and after this PR I get a list with a single item `[18446744073709551616]` on linux64, python 3.8. What am I missing?

BvB93 commented:
> I think this breaks `np.trim_zeros([2**64, 0])`.

Overflow shouldn't be an issue actually, as `np.array()` will automatically default to an alternative dtype if an integer is too large. Note that all below-mentioned data types work just fine with `ndarray.astype(bool)`.

``` python
In [1]: import numpy as np                                                                                             

In [2]: np.asanyarray(2**62)                                                                                                
Out[2]: array(4611686018427387904)

In [3]: np.asanyarray(2**63)                                                                                                
Out[3]: array(9223372036854775808, dtype=uint64)

In [4]: np.asanyarray(2**64)                                                                                                
Out[4]: array(18446744073709551616, dtype=object)
```

mhvk commented:
Note that this breaks astropy [1] as we can compare Quantities to 0 but cannot turn them into bool (not clear that '0 m' should be `False`, after all). While it isn't really a big deal since we can override with `__array_function__`, I wondered if it would be OK to have a follow-up PR to change the `astype(bool)` to comparing with 0 -- it not only is closer to what the name promises, but also turns out to be faster (found this out while trying to see how much worse it would be!):
```
In [4]: a = np.hstack([np.zeros(100_000), np.random.uniform(size=100_000), np.zeros(100_000)])

In [5]: %timeit b = a.astype(bool)
288 µs ± 3.61 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)

In [6]: %timeit b = a != 0
97.6 µs ± 185 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```

[1] https://github.com/astropy/astropy/issues/10638

charris commented:
@mhvk I get about that same times for both methods in master.
```
In [2]: %timeit b = a.astype(bool)                                              
77.9 µs ± 85.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)

In [3]: %timeit b = a != 0                                                      
70.2 µs ± 149 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
```
Wonder what's up with that. Do you compile your own NumPy?

mhvk commented:
That's weird! I was running in a virtual environment with numpy installed from source (as I was triaging the astropy failure), but checking my vanilla debian testing numpy 1.18.4, I get essentially the same timings as those I posted. (For both, python 3.8.5, compilation likely gcc 9.3, not sure what else would be informative.)

mhvk commented:
p.s. It is probably handier if @BvB93 makes a PR, though in principle I can do it too. Note that my suggestion does change test results for the `object` array (no entry of which will compare True to 0) and also it will not fail one of the two deprecation test cases (the one with `str`). Since in both cases that just reverts to the behaviour prior to this PR, I think that is, if anything, a benefit.

seberg commented:
Not that it matters, but I missed that this never tested the case where no stripping occurs (i.e. I was not clear enough as to what I meant before). I think basically the old code was optimized for that, while the new code will cast the full array even though almost nothing happens.

charris commented:
Python 3.8.5 here also, gcc 10.2.1. For some reason my 2013 I5 setup beats a lot of modern machines. Apparently [Falcon Northwest](https://www.falcon-nw.com/) is as good as they claim :) Anyway CFLAGS are
```
-Wall -Wstrict-prototypes -O3 -pipe -fomit-frame-pointer -fno-strict-aliasing -Wmaybe-uninitialized -Wdeprecated-declarations -march=native
```
in case it matters.

mhvk commented:
My timings go down to 75 us and 240 us if I plug in power... I tried my office computer too (not very new Xeon, same Debian testing), and it gives 108, 257 us. Suggests perhaps some compiler/chipset optimization done for one case but not the other.

@seberg - the ideal solution might be to use the `.first` reduction type ufunc method that @ahaldane suggested a long time ago: https://github.com/numpy/numpy/pull/8528#issuecomment-365358119

BvB93 commented:
I'm also seeing a noticable speedup here when running the new benchmark, 
which in and of itself would, in my opinion, make this a worthwhile change. 

I'll create a follow up pull request in a bit, as @mhvk suggested.

Before
-------
```bash
>>> python runtests.py --bench bench_trim_zeros
...
[100.00%] ··· ===================== ============ ========== ==========
              --                                   size               
              --------------------- ----------------------------------
                      dtype             3000       30000      300000  
              ===================== ============ ========== ==========
                  dtype('int64')      9.05±1μs    35.8±5μs   315±40μs 
                 dtype('float64')    15.1±0.5μs   56.1±6μs   490±90μs 
               dtype('complex128')    20.9±3μs    86.3±7μs   891±60μs 
                  dtype('bool')      10.5±0.6μs   28.4±1μs   205±5μs  
              ===================== ============ ========== ==========
```

After
-----
```bash
[100.00%] ··· ===================== ============ ========== ===========
              --                                    size               
              --------------------- -----------------------------------
                      dtype             3000       30000       300000  
              ===================== ============ ========== ===========
                  dtype('int64')     6.24±0.7μs   22.8±2μs    174±7μs  
                 dtype('float64')     11.9±1μs    40.2±4μs    321±40μs 
               dtype('complex128')    15.3±2μs    77.6±9μs   837±100μs 
                  dtype('bool')       15.0±2μs    63.6±6μs    537±30μs 
              ===================== ============ ========== ===========
```

BvB93 commented:
I just created a followup at https://github.com/numpy/numpy/pull/17058.

eric-wieser commented:
> I'm also seeing a noticable speedup here when running the new benchmark,
> which in and of itself would, in my opinion, make this a worthwhile change.

Performance usually isn't in itself a sufficient motivation for a behavior change!

seberg commented:
@mhvk sorry, it only sunk in now that you said its not clear that `0 m` should be considered `False` when casting to bool. Is there any particular reason for that? That basically means a quantity (scalar), does not define `bool(0 m)` at all (I see that this is specfically written like that).
I am a bit surprised by that, although I guess there are a few tricky points where it is indeed not clearly defined (e.g. temperature could only use absolute 0).  So I guess the reason for the choice is that it does not work for units that scale multiplicative? (I can't think of what that is called).
Its interesting, because in that case I would assume comparison with 0 should also fail:
```
c = Quantity(0, "Celsius")
print(c == 0)  # True
c = Quantity(1, "Celsius")
print(c == 1)  # False
```
Its curious how comparison with 0 is special cased, but `.astype(bool)` discouraged, on first sight I would almost expect the opposite `quantity == 0` is always `False` or an error, but `bool(quantity)` works for meters (but maybe not Celsius).

mhvk commented:
@eric-wieser - but performance improvement plus *not* changing behaviour compared to released code is surely nice!

@seberg - agreed that allowing comparison with 0 is also not so obvious. And indeed we do a full unit check if the 0 has a unit attached. Only 0 without unit is special-cased (as are `inf` and `nan`), in allowing them to have an arbitrary unit. It is mostly because for most units, `q = 0` or, e.g., `q > 0` has very obvious meaning (and, frankly, because originally it meant a few numpy functions such as `np.sinc` -- which does `np.where(x==0, ...)` 'just worked' - see https://github.com/astropy/astropy/pull/1254). I guess for me it boils down to thinking of quantities as often representing relative measurements, in which case it becomes possible to see how a quantity can represent "sameness" but still remains tricky to infer any "truthiness".

seberg commented:
@mhvk I suppose my point would be that if you can argue about `== 0` being useful, then it seems that in those cases thruthiness may also be usefully defined?  I agree that it probably is a swamp though.

@eric-wieser actually, `== 0` seems like the smaller behaviour change... `None == 0` is False and that was used previously, nonzero is the possibly the more *reasonable* behaviour, but probably the behaviour switch was the original PR, this actually reverts it back to the previous behaviour!

eric-wieser commented:
Ah, I'd missed that we already made the behavior change.

mhvk commented:
Agreed, thruthiness could be defined -- I guess it boils down to never having had the need...  And it remains tricky anyway, with always the option to go for `q.size > 0`, in analogy with a list. But this is unsolved for numpy as well - I wish `bool(array)` was actually defined, at least for boolean arrays....

seberg commented:
True, in any case, the discussion is a bit moot, since we would be discussing changing the behaviour away from the current `== 0` and I am not sure there is a reason for that now.  Yeah, since there is no clear quantity scalar, there is no clear way to resolve the truthiness issue (just as we have no clear numpy scalars, even though we have scalars ;)).

BvB93 commented:
After seeing Eric's comment about `np.count_nonzero()` (https://github.com/numpy/numpy/pull/17058#issuecomment-672000222) it might actually be possible to kill two birds with one stone here. Adding the lines below would add support for more dtypes (`character` & `void`) _and_ fix the `astropy.Quantity` issue all in one go.

Any thoughts?

``` python
arr_any = np.asanyarray(...)
try:
    arr = arr_any.astype(bool, copy=False)
except (TypeError, ValueError):
    arr = arr_any != arr_any.dtype.type()
```

seberg commented:
Lets not be tempted to a try/except, its too hard to grasp all the possible ways that can go wrong.

The current (1.19.x) behaviour is to use `element == 0.`, the new behaviour effectively uses `bool(element)`. That is a change in behaviour which trips astropy – for whatever reasons.

That behaviour is well defined (maybe weird, but seems also not super bad). We could discuss moving the function to mean `stip_falsy`, in which case I am actually not sure breaking astropy matters. If it does it would be an argument against changing the meaning.  But trying to do some in-between thing seems very muddy waters to me.

But for starters, I mostly see an unintended behaviour change, and whether that is actually better behaviour can be discussed, but it would be good to have use-cases.

BvB93 commented:
In that case I'd say it would be the simplest to revert to the old 1.19 `element == 0` behavior.

The main purpose of this pull request was to provide a speed up to `trim_zeros()` and, quite frankly, I did not expect switching to the `bool(element)` approach could realistically break pre-existing pieces of code. But it did.

mhvk commented:
@BvB93 - thanks for the new PR. Don't blame you for not expecting the break - it is only because when implementing `__array_function__` I wanted to be complete that it is even tested... But I guess one often learns something new: I really had not expected the comparison to be even equally fast, let alone faster!
